### PR TITLE
base.py: Make ssh/scp use additional ssh key

### PR DESCRIPTION
**Pull request by haoxli**
In order to enable test service when multiple users login on Windows, ssh/scp need to use a same private key to access backup server for each user.

### CODE REVIEW

**Comment by haoxli** [comment]
@gyagp  PTAL, thansk! Next I will add a parameter in gputest to control whether use this ssh key when connecting to backup server, and finally enable this parameter in our service for Windows.

The same change will happen in ort test.

**Comment by haoxli** [comment]
@gyagp  PTAL!

File: base.py [hunk 1]
```diff
@@ -1185,22 +1185,36 @@ def get_backup_dir(backup_dir, rev):
...
-    def remotify_cmd(server, cmd, extra_arg=''):
-        new_cmd = 'ssh'
+    # options: [-46AaCfGgKkMNnqsTtVvXxYy] [-B bind_interface]
+    #         [-b bind_address] [-c cipher_spec] [-D [bind_address:]port]
+    #         [-E log_file] [-e escape_char] [-F configfile] [-I pkcs11]
+    #         [-i identity_file] [-J [user@]host[:port]] [-L address]
+    #         [-l login_name] [-m mac_spec] [-O ctl_cmd] [-o option] [-p port]
+    #         [-Q query_option] [-R address] [-S ctl_path] [-W host:port]
+    #         [-w local_tun[:remote_tun]]
+    def ssh_cmd(server, cmd='', options=''):
```
Comment (gyagp) [inline base.py]: We don't need options as parameter, for simplicity.

File: base.py [hunk 2]
```diff
@@ -1185,22 +1185,36 @@ def get_backup_dir(backup_dir, rev):
...
-        new_cmd += ' "%s"' % cmd
-        return new_cmd
+            options += ' -x'
+        return f'ssh {options} wp@{server} {cmd}'
+
+    @staticmethod
+    # options: [-346ABCpqrTv] [-c cipher] [-F ssh_config] [-i identity_file]
+    #         [-J destination] [-l limit] [-o ssh_option] [-P port]
+    #         [-S program]
+    def scp_cmd(source, target, options=''):
```
Comment (gyagp) [inline base.py]: We don't need options as parameter, for simplicity.

Comment (gyagp) [inline base.py]: source -> src, target -> dest

File: base.py [hunk 3]
```diff
@@ -1229,7 +1239,8 @@ def get_server_backup(relative_path, rev='latest'):
         local_backup_dir = '%s/%s' % (Util.BACKUP_DIR, relative_path)
         Util.ensure_dir(local_backup_dir)
         if not os.path.exists('%s/%s' % (local_backup_dir, rev_name)) and not os.path.exists('%s/%s' % (local_backup_dir, rev_file)):
-            Util.execute('scp wp@%s:/workspace/backup/%s/%s/%s %s' % (Util.BACKUP_SERVER, Util.HOST_OS, relative_path, rev_file, local_backup_dir))
+            cmd = Util.scp_cmd(f'wp@{Util.BACKUP_SERVER}:/workspace/backup/{Util.HOST_OS}/{relative_path}/{rev_file}', f'{local_backup_dir}')
```
Comment (gyagp) [inline base.py]: f'{local_backup_dir}' -> local_backup_dir

File: base.py [hunk 4]
```diff
@@ -1323,6 +1334,11 @@ def get_local_backup(relative_path, rev='latest'):
     GNP_SCRIPT =  format_slash.__func__('%s/misc/gnp.py' % PROJECT_TOOLKIT_DIR)
     MESA_SCRIPT = format_slash.__func__('%s/misc/mesa.py' % PROJECT_TOOLKIT_DIR)
 
+    # We cannot ensure that all users on a host have put their ssh public keys to backup server,
```
Comment (gyagp) [inline base.py]: We don't need to mention backup server here.

File: base.py [hunk 5]
```diff
@@ -1323,6 +1334,11 @@ def get_local_backup(relative_path, rev='latest'):
...
+    # so use a shared key instead of personal key to allow the service to always access the server,
+    # no matter which user logs in.
+    SSH_KEY = format_slash.__func__('%s/ssh_key/id_rsa' % PROJECT_DIR)
```
Comment (gyagp) [inline base.py]: I think we can put it in TOOL_DIR, as it's not a project.

Comment (haoxli) [reply base.py]: The projects using util will have different TOOL_DIRs, we need to put the keys to each project, but I just want to keep them in one place on a host.
Or we upload the keys to util/tool, which would violate security policy, this is a public repo and we'd exposed some server names in it.

If PROJECT_DIR is not a good place, how about PROJECT_TOOLKIT_DIR where it's really used.

Comment (haoxli) [reply base.py]: One more point, other projects may also require the key, such as ort test. The path for the keys should be as universal as possible.

Comment (gyagp) [reply base.py]: You're right. I was thinking of some folder in "d:\workspace\tools". Folder project supposes to host all the git repos. I did want to bring in a folder like this, but I changed the mind later on to check in the small tools to repo util to save maintenance cost.

Comment (gyagp) [reply base.py]: Can we just use repo work, and put private key at config/common/? In this way, we don't need any extra setup on various machines. I really hope we can automatically maintain all the machines, including the introduction of new machines.

Comment (haoxli) [reply base.py]: Ok, I will upload the key to work/config/common, then automatically copy to each user's .ssh directory, we will use the key copied there.
Also defined a `SSH_DIR` to check whether the path exists when copying the file.

**Review by gyagp — APPROVED** [review]
LGTM. Thanks for working on this!

**Pull request merged**